VM: Added Implementations of Variable

enghqii opened the pull request:
1. VarDefStatement의 생성자를 yacc의 rule에 맞게 수정했습니다.
2. # 86 에 언급한대로 context 를 구현했습니다. (테이블 추가, BlockBegin/End)
3. 변수와 관련 있는 expression 들의 내용을 채워넣었습니다. (VariableExpression, AssignExpression)
4. 위에서 수정, 생성한 클래스에 맞춰 Yacc action코드를 추가했습니다. 
5. 테스트 코드들을 추가했습니다.

( cout으로 출력하던 로그도 지웠습니다)

enghqii committed:
Stmt: Updated 'VarDefStatement's constructors

VarDefStatement의 생성자를 yacc에서 사용할 수 있게 확장하고, Execute도 추가했습니다 (아직 변수 테이블에 Insert하지 않음)

enghqii committed:
Context: Added BlockBegin/End

Context
context chaining을 구현했습니다. Context내부에 테이블들을 집어넣고, Find함수들을 추가했습니다. BlockBegin, BlockEnd함수들을 구현했습니다. VM에서 DataTypeTable을 이용한 생성, Child context를 만들기 위해 context* 를 이용한 생성, 두개의 생성자들을 구현했습니다.

Block, VirtualMachine, VarDefStatement
구현된 Context::BlockBegin/End 에 대한 구현코드를 추가 했습니다.

enghqii committed:
Expr: VarExpr, AssignExpr implemented

VariableExpression::Evaluate() 를 구현했습니다.

AssignExpr를 구현하기 위해 VarExpr::getVariable(Context)를 추가하고, AssignExpr의 Evaluate를 구현했습니다.

enghqii committed:
Yacc: Test: Updated action code for variable

program, AssignExpression, VariableExpression, var_def_statement에 대한 액션 코드 추가

expression non-terminal에서 assign부분이 같이 있으면 한눈에 보기 힘들어서 예전 버전 처럼 assign operator는 따로 빼냈습니다.

'program' rule의 action에서 SetDataTypeTableToVM을 할때, VM 인스턴스를 넘겨주기 위해 ParserContext 의 생성자 부분에서 VM을 만들어서 넣어줍니다.

enghqii committed:
Test: For test added

namhyung commented:
Block::Execute()에서 continue/break 등의 검사는 child context로 해야하지 않을까요?

그리고 커밋 메시지에 좀 더 신경써 주세요. 제목이 너무 길면 보기 좋지 않고, 내용이 없으면 나중에 알기 어렵습니다. 가령 fix xxx 라고만 하면 그게 왜 문제가 되는지 언제부터 문제가 된건지 모르겠지요.

enghqii committed:
VM: Changed Context's Jump flags into pointers

1. Context 에 있는 점프 플래그들을 부모-자식관계에서도 하나의 변수를 모든 Context에 공유하기 위해 (DataTypeTable처럼) 포인터로 수정했습니다. break, continue는 ForStatement를 나올때 플래그를 false로 꺼줍니다. 이후에 return플래그는, FunctionExp가 종료될때 꺼줍니다.

2. Exit 플래그를 지웠습니다. ExitStatement에서 바로 exit(int) 함수를 호출합니다.
